lib/makeCmdTasks.js

```diff
@@ -27,25 +27,42 @@ module.exports = async function makeCmdTasks({ commands, files, gitDir, shell })
     const resolvedArray = Array.isArray(resolved) ? resolved : [resolved] // Wrap non-array command as array
 
     for (const command of resolvedArray) {
-      let title = isFn ? '[Function]' : command
+      let title = isFn ? '[Function] ' : ''
 
-      if (isFn) {
-        // If the function linter didn't return string | string[]  it won't work
-        // Do the validation here instead of `validateConfig` to skip evaluating the function multiple times
-        if (typeof command !== 'string') {
-          throw new Error(
-            createError(
-              title,
-              'Function task should return a string or an array of strings',
-              resolved
-            )
+      // If the function linter didn't return string | string[]  it won't work
+      // Do the validation here instead of `validateConfig` to skip evaluating the function multiple times
+      if (isFn && typeof command !== 'string') {
+        throw new Error(
+          createError(
+            title.trim(),
+            'Function task should return a string or an array of strings',
+            resolved
           )
+        )
+      }
+
+      const words = command.trim().split(' ') // Tokenize the command into words
+
+      if (isFn) {
+        const indexOfFile = words.findIndex(word => files.includes(word)) // index of first file in the command
+        if (indexOfFile >= 0) {
+          words.splice(indexOfFile) // Strip staged files from command
         }
+      }
 
-        const [startOfFn] = command.split(' ')
-        title += ` ${startOfFn} ...` // Append function name, like `[Function] eslint ...`
+      // For the title, strip package runner commands (npm/npx/yarn) from the beginning
+      if (words.length > 0 && ['npm', 'npx', 'yarn'].includes(words[0])) {
+        let idx = 1
+        while (idx < words.length && (words[idx] === '' || words[idx].startsWith('-'))) {
+          idx += 1 // Remove any options to the package runner command
+        }
+        if (idx < words.length) {
+          words.splice(0, idx) // Make sure we don't strip the entire command
+        }
```

Comment:
Not sure about this. How will it format something like `(files) => npm run test -- --update ${files.join(' ')}`

Reply:
It'll format it as `run test -- --update`. Do you think it makes sense to special-case the `run` command? Or maybe in general it is going to be too difficult to properly handle package runner commands, without overcomplicating things?

Reply:
I think it’s best to leave out stripping those. I’d imagine there’s always some reason to run via npm/yarn instead of directly, so better not hide it.

Reply:
Sounds good, I'll update the PR.